feat: append variable qualities to dpr srcsets

sherwinski opened the pull request:
This PR automatically appends a `q` parameter mapped to each `dpr` parameter when generating a [fixed-size srcset](https://github.com/imgix/imgix-rb#fixed-image-rendering). This technique is commonly used to compensate for the increased filesize of high-DPR images. Since high-DPR images are displayed at a higher pixel density on devices, image quality can be lowered to reduce overall filesize without sacrificing perceived visual quality. For more information and examples of this technique in action, see [this blog post](https://blog.imgix.com/2016/03/30/dpr-quality).
This behavior will respect any overriding `q` value passed in as a parameter. Additionally, it can be disabled altogether by passing `options: { disable_variable_qualities: true }` to `Imgix:Path#to_srcset`.

The following map of `q` to `dpr` pairs will be generated:

`dpr` | `q`
--- |---
1 | 75 
2 | 50 
3 | 35 
4 | 23 
5 | 20 

This behavior specifically occurs when a [fixed-size image](https://github.com/imgix/imgix-rb#fixed-image-rendering) is rendered, for example:

```rb
srcset = Imgix::Client.new(host: 'testing.imgix.net', include_library_param: false)
.path('image.jpg')
.to_srcset(w:100)
```
will generate the following srcset:

```html
https://testing.imgix.net/image.jpg?w=100&dpr=1&q=75 1x,
https://testing.imgix.net/image.jpg?w=100&dpr=2&q=50 2x,
https://testing.imgix.net/image.jpg?w=100&dpr=3&q=35 3x,
https://testing.imgix.net/image.jpg?w=100&dpr=4&q=23 4x,
https://testing.imgix.net/image.jpg?w=100&dpr=5&q=20 5x
```

Commit:
feat: add variable qualities to srcset pairs

Commit:
test: add tests for variable qualities and override

Commit:
add srcset capability to disable variable qualities

Commit:
test: disable variable qualities test cases

Commit:
docs: add documentation and example

sherwinski commented on lib/imgix/path.rb:
```diff
@@ -154,12 +154,23 @@ def build_srcset_pairs(options:, params:)
+    def build_dpr_srcset(options:, params:)
       srcset = ''
+
+      disable_variable_qualities = options['disable_variable_qualities'.to_sym] || false
```
Extremely open to shorter name suggestions here for disabling variable qualities.

jayeb replied:
`disable_variable_quality` is *slightly* shorter... ¯\_(ツ)_/¯

sherwinski replied:
I'll take it

jayeb commented on lib/imgix/path.rb:
```diff
@@ -154,12 +154,23 @@ def build_srcset_pairs(options:, params:)
+        unless disable_variable_qualities
+          params['q'.to_sym] = quality || DPR_QUALITY[index]
+        end
+
```
Would this be a little cleaner if `DPR_QUALITY` was a map instead of an array, with the target ratios as the keys?

sherwinski replied:
It's definitely more self-documenting, but then `index` won't align with `DPR_QUALITY` and `target_ratios` since we access the hash by the key and not an index like an array. So it would look like this:

```rb
# imgix.rb
DPR_QUALITY = {
    1 => 75, 
    2 => 50,
    3 => 35,
    4 => 23,
    5 => 20
  }

# path.rb
params['q'.to_sym] = quality || DPR_QUALITY[index + 1]
```

I don't think that means we shouldn't do it though. It's just a bit of a  trade-off in terms of "cleaner."

jayeb replied:
Or you ditch the `index` thing entirely and go back to `for ratio in target_ratios do` above. then you could do `params['q'.to_sym] = quality || DPR_QUALITY[ratio]`, right?

sherwinski replied:
You're absolutely right 😂

jayeb commented on test/units/srcset_test.rb:
```diff
@@ -68,7 +68,34 @@ def test_srcset_signs_urls
+                assert(not(src.include? "q="))
+            }
+        end
+
```
Can we add one more test to make sure an overriding `q` param is still respected even if `disable_variable_qualities` is false?

Commit:
Address PR feedback:
- restructure DPR_QUALITY as a hash
- rename parameter to 'disable_variable_quality'
- add test for overriding q when variable quality is disabled

Commit:
cleaner dpr srcset building
